cpp/src/reductions/reduction_operators.cuh

```diff
@@ -1,58 +1,183 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CUDF_REDUCTION_OPERATORS_CUH
+#define CUDF_REDUCTION_OPERATORS_CUH
+
 #include <cudf/cudf.h>
 #include <utilities/cudf_utils.h>
 #include <utilities/wrapper_types.hpp>
 #include <utilities/error_utils.hpp>
+#include <iterator/iterator.cuh>
+
+#include <utilities/device_operators.cuh>
 
-#include <utilities/device_atomics.cuh>
+#include <cmath>
 
 namespace cudf {
 namespace reductions {
 
```

Comment:
Perhaps another `namespace op` in here so you don't end up with duplicate names for the reduction functions and the operator structs. e.g. `cudf::reductions:mean` and `cudf::reductions::op::mean`? Or you could name them all `op_mean`, `op_sum`, etc.

Reply:
That's good idea, 
I will use `cudf::reductions:mean` as function and `cudf::reductions::op::mean` as operator struct.